File: app/utils/string_utils.py
import unicodedata
import re

def remover_caracteres_especiais(texto: str) -> str:
    """Remove caracteres especiais (acentos, cedilhas) de uma string."""
    # Normaliza o texto para decompor os caracteres acentuados
    texto_normalizado = unicodedata.normalize('NFKD', texto)
    # Remove caracteres que não são ASCII (como acentos e cedilhas)
    return ''.join([c for c in texto_normalizado if unicodedata.category(c) != 'Mn'])

def normalizar_nome_cidade(cidade: str) -> str:
    """
    Normaliza o nome de uma cidade para busca no banco de dados.
    Remove pontuação, normaliza espaços e trata variações comuns.
    """
    if not cidade:
        return ""
    
    # Converte para maiúsculas e remove espaços extras
    cidade = cidade.upper().strip()
    
    # Normaliza espaços múltiplos
    cidade = re.sub(r'\s+', ' ', cidade)
    
    return cidade.strip()

def normalizar_nome_cidade_sem_pontuacao(cidade: str) -> str:
    """
    Normaliza o nome de uma cidade removendo pontuação para busca flexível.
    """
    if not cidade:
        return ""
    
    # Converte para maiúsculas e remove espaços extras
    cidade = cidade.upper().strip()
    
    # Remove pontuação (apóstrofos, hífens, pontos, etc.)
    cidade = re.sub(r'[^\w\s]', '', cidade)
    
    # Normaliza espaços múltiplos
    cidade = re.sub(r'\s+', ' ', cidade)
    
    # Trata variações comuns
    cidade = cidade.replace("DO OESTE", "DOESTE")
    cidade = cidade.replace("D OESTE", "DOESTE")
    cidade = cidade.replace("DA PRAIA", "DAPRAIA")
    cidade = cidade.replace("D PRAIA", "DAPRAIA")
    cidade = cidade.replace("DE GOIAS", "DEGOIAS")
    cidade = cidade.replace("D GOIAS", "DEGOIAS")
    
    return cidade.strip()

def normalizar_nome_cidade_sem_acentos(cidade: str) -> str:
    """
    Normaliza o nome de uma cidade removendo acentos para busca flexível.
    """
    if not cidade:
        return ""
    
    # Converte para maiúsculas e remove espaços extras
    cidade = cidade.upper().strip()
    
    # Remove acentos usando unicodedata
    cidade_sem_acentos = remover_caracteres_especiais(cidade)
    
    # Normaliza espaços múltiplos
    cidade_sem_acentos = re.sub(r'\s+', ' ', cidade_sem_acentos)
    
    return cidade_sem_acentos.strip()

def normalizar_nome_cidade_completa(cidade: str) -> str:
    """
    Normaliza o nome de uma cidade removendo acentos e pontuação para busca flexível.
    """
    if not cidade:
        return ""
    
    # Converte para maiúsculas e remove espaços extras
    cidade = cidade.upper().strip()
    
    # Remove acentos
    cidade = remover_caracteres_especiais(cidade)
    
    # Remove pontuação
    cidade = re.sub(r'[^\w\s]', '', cidade)
    
    # Normaliza espaços múltiplos
    cidade = re.sub(r'\s+', ' ', cidade)
    
    # Trata variações comuns
    cidade = cidade.replace("DO OESTE", "DOESTE")
    cidade = cidade.replace("D OESTE", "DOESTE")
    cidade = cidade.replace("DA PRAIA", "DAPRAIA")
    cidade = cidade.replace("D PRAIA", "DAPRAIA")
    cidade = cidade.replace("DE GOIAS", "DEGOIAS")
    cidade = cidade.replace("D GOIAS", "DEGOIAS")
    
    # Trata variações comuns de nomes de cidades
    cidade = cidade.replace("POXOREU", "POXOREO")  # Poxoréu -> Poxoreo
    cidade = cidade.replace("CUIABA", "CUIABÁ")    # Cuiaba -> Cuiabá
    cidade = cidade.replace("VARZEA", "VÁRZEA")    # Varzea -> Várzea
    
    return cidade.strip()

def gerar_padroes_busca_cidade(cidade: str) -> list:
    """
    Gera uma lista de padrões de busca para uma cidade,
    considerando variações com pontuação, acentos e sem pontuação.
    """
    if not cidade:
        return []
    
    padroes = []
    
    # Padrão original (mantém apóstrofos, pontuação e acentos)
    cidade_original = cidade.upper().strip()
    padroes.append(cidade_original)
    
    # Padrão sem acentos (mantém pontuação)
    cidade_sem_acentos = normalizar_nome_cidade_sem_acentos(cidade)
    padroes.append(cidade_sem_acentos)
    
    # Padrão normalizado (sem pontuação, mantém acentos)
    cidade_norm = normalizar_nome_cidade_sem_pontuacao(cidade)
    padroes.append(cidade_norm)
    
    # Padrão completamente normalizado (sem acentos e sem pontuação)
    cidade_completa = normalizar_nome_cidade_completa(cidade)
    padroes.append(cidade_completa)
    
    # Se contém "D'OESTE", criar variações
    if "D'OESTE" in cidade_original:
        nome_base = cidade_original.replace("D'OESTE", "").strip()
        padroes.extend([
            f"{nome_base} D'OESTE",
            f"{nome_base} DO OESTE",
            f"{nome_base} D OESTE",
            f"{nome_base}%OESTE"
        ])
    
    # Se contém "DO OESTE", criar variações
    elif "DO OESTE" in cidade_original:
        nome_base = cidade_original.replace("DO OESTE", "").strip()
        padroes.extend([
            f"{nome_base} DO OESTE",
            f"{nome_base} D'OESTE",
            f"{nome_base} D OESTE",
            f"{nome_base}%OESTE"
        ])
    
    # Se contém "DA PRAIA", criar variações
    elif "DA PRAIA" in cidade_original:
        nome_base = cidade_original.replace("DA PRAIA", "").strip()
        padroes.extend([
            f"{nome_base} DA PRAIA",
            f"{nome_base} D PRAIA",
            f"{nome_base}%PRAIA"
        ])
    
    # Se contém "DE GOIAS", criar variações
    elif "DE GOIAS" in cidade_original:
        nome_base = cidade_original.replace("DE GOIAS", "").strip()
        padroes.extend([
            f"{nome_base} DE GOIAS",
            f"{nome_base} D GOIAS",
            f"{nome_base}%GOIAS"
        ])
    
    # Se contém "POXORÉU", criar variações
    elif "POXORÉU" in cidade_original:
        padroes.extend([
            "POXOREO",
            "POXORÉU",
            "%POXOR%"
        ])
    
    # Se contém "POXOREU", criar variações
    elif "POXOREU" in cidade_original:
        padroes.extend([
            "POXOREO",
            "POXORÉU",
            "%POXOR%"
        ])
    
    # Padrões com wildcard para busca flexível
    padroes.extend([
        f"%{cidade_original}%",      # Com acentos e pontuação
        f"%{cidade_sem_acentos}%",   # Sem acentos, com pontuação
        f"%{cidade_norm}%",          # Com acentos, sem pontuação
        f"%{cidade_completa}%"       # Sem acentos e sem pontuação
    ])
    
    # Remove duplicatas mantendo a ordem
    padroes_unicos = []
    for padrao in padroes:
        if padrao not in padroes_unicos:
            padroes_unicos.append(padrao)
    
    return padroes_unicos 

File: src/utils/stringUtils.ts
/**
 * Utilitários para normalização de strings
 * Funções globais para uso em todo o projeto
 */

// ============================================================================
// FUNÇÕES BÁSICAS DE NORMALIZAÇÃO
// ============================================================================

/**
 * Remove acentos de uma string
 * @param texto - String com acentos
 * @returns String sem acentos
 */
export const removeAcentos = (texto: string): string => {
  return texto.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
};

/**
 * Normaliza string para minúsculas (remove acentos, converte para minúsculas)
 * @param texto - String a ser normalizada
 * @returns String normalizada em minúsculas
 */
export const normalizarString = (texto: string): string => {
  return removeAcentos(texto).toLowerCase().trim();
};

/**
 * Normaliza string para maiúsculas (remove acentos, converte para maiúsculas)
 * @param texto - String a ser normalizada
 * @returns String normalizada em maiúsculas
 */
export const normalizarStringMaiuscula = (texto: string): string => {
  return removeAcentos(texto).toUpperCase().trim();
};

// ============================================================================
// FUNÇÕES ESPECÍFICAS PARA CIDADES
// ============================================================================

/**
 * Normaliza nome de cidade para corresponder ao formato do banco de dados
 * Remove acentos, caracteres especiais, converte para maiúsculas
 * e aplica regras específicas para cidades
 * 
 * @param nomeCidade - Nome da cidade a ser normalizado
 * @returns Nome da cidade normalizado
 */
export const normalizarNomeCidade = (nomeCidade: string): string => {
  let normalizada = nomeCidade
    // Remove acentos
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    // Remove caracteres especiais exceto espaços
    .replace(/[^a-zA-Z0-9\s]/g, ' ')
    // Converte para maiúsculas
    .toUpperCase()
    // Remove espaços extras
    .trim()
    // Substitui "DO" por "D" para corresponder ao banco
    .replace(/\bDO\b/g, 'D')
    // Remove espaços duplicados
    .replace(/\s+/g, ' ');
  
  // Tratamentos específicos para casos especiais
  normalizada = aplicarTratamentosEspeciais(normalizada);
  
  return normalizada;
};

/**
 * Normaliza nome de cidade para envio à API
 * Alias para normalizarNomeCidade para manter compatibilidade
 * 
 * @param nomeCidade - Nome da cidade a ser normalizado
 * @returns Nome da cidade normalizado para API
 */
export const normalizarNomeCidadeParaAPI = (nomeCidade: string): string => {
  return normalizarNomeCidade(nomeCidade);
};

/**
 * Aplica tratamentos específicos para casos especiais de cidades
 * @param nomeNormalizado - Nome já normalizado
 * @returns Nome com tratamentos especiais aplicados
 */
const aplicarTratamentosEspeciais = (nomeNormalizado: string): string => {
  // Mapeamento de casos especiais
  const casosEspeciais: Record<string, string> = {
    'CURVELANDIA': 'CUVERLANDIA',
    // Adicione novos casos especiais aqui
    // 'NOME_ORIGINAL': 'NOME_CORRIGIDO',
  };
  
  return casosEspeciais[nomeNormalizado] || nomeNormalizado;
};

// ============================================================================
// FUNÇÕES DE COMPARAÇÃO
// ============================================================================

/**
 * Compara dois nomes de cidades normalizando ambos
 * @param nome1 - Primeiro nome de cidade
 * @param nome2 - Segundo nome de cidade
 * @returns true se os nomes normalizados forem iguais
 */
export const compararNomesCidades = (nome1: string, nome2: string): boolean => {
  const normalizado1 = normalizarNomeCidade(nome1);
  const normalizado2 = normalizarNomeCidade(nome2);
  return normalizado1 === normalizado2;
};

/**
 * Encontra uma cidade em uma lista usando normalização
 * @param cidadeBusca - Cidade a ser buscada
 * @param listaCidades - Lista de cidades para buscar
 * @returns A cidade encontrada ou null
 */
export const encontrarCidade = (cidadeBusca: string, listaCidades: string[]): string | null => {
  const cidadeNormalizada = normalizarNomeCidade(cidadeBusca);
  
  for (const cidade of listaCidades) {
    if (normalizarNomeCidade(cidade) === cidadeNormalizada) {
      return cidade;
    }
  }
  
  return null;
};

// ============================================================================
// FUNÇÕES DE VALIDAÇÃO
// ============================================================================

/**
 * Verifica se um nome de cidade é válido
 * @param nomeCidade - Nome da cidade a ser validado
 * @returns true se o nome for válido
 */
export const validarNomeCidade = (nomeCidade: string): boolean => {
  if (!nomeCidade || typeof nomeCidade !== 'string') {
    return false;
  }
  
  const nomeLimpo = nomeCidade.trim();
  return nomeLimpo.length > 0 && nomeLimpo.length <= 100;
};

/**
 * Verifica se uma cidade existe em uma lista
 * @param cidade - Cidade a ser verificada
 * @param listaCidades - Lista de cidades
 * @returns true se a cidade existir na lista
 */
export const cidadeExiste = (cidade: string, listaCidades: string[]): boolean => {
  return encontrarCidade(cidade, listaCidades) !== null;
};

// ============================================================================
// FUNÇÕES DE UTILIDADE
// ============================================================================

/**
 * Normaliza uma lista de cidades
 * @param cidades - Lista de cidades a serem normalizadas
 * @returns Lista de cidades normalizadas
 */
export const normalizarListaCidades = (cidades: string[]): string[] => {
  return cidades.map(cidade => normalizarNomeCidade(cidade));
};

/**
 * Remove duplicatas de uma lista de cidades usando normalização
 * @param cidades - Lista de cidades
 * @returns Lista sem duplicatas
 */
export const removerDuplicatasCidades = (cidades: string[]): string[] => {
  const cidadesUnicas = new Set<string>();
  const resultado: string[] = [];
  
  for (const cidade of cidades) {
    const normalizada = normalizarNomeCidade(cidade);
    if (!cidadesUnicas.has(normalizada)) {
      cidadesUnicas.add(normalizada);
      resultado.push(cidade);
    }
  }
  
  return resultado;
};

/**
 * Filtra cidades que começam com um prefixo (case-insensitive)
 * @param cidades - Lista de cidades
 * @param prefixo - Prefixo para filtrar
 * @returns Lista filtrada
 */
export const filtrarCidadesPorPrefixo = (cidades: string[], prefixo: string): string[] => {
  const prefixoNormalizado = normalizarString(prefixo);
  
  return cidades.filter(cidade => {
    const cidadeNormalizada = normalizarString(cidade);
    return cidadeNormalizada.startsWith(prefixoNormalizado);
  });
};

// ============================================================================
// EXPORTAÇÕES PARA COMPATIBILIDADE
// ============================================================================

// Re-exportar funções antigas para manter compatibilidade
export {
  normalizarNomeCidade as normalizarCidade,
  normalizarNomeCidadeParaAPI as normalizarCidadeParaAPI,
};
